xarray/backends/api.py

```diff
@@ -1118,7 +1118,8 @@ def to_zarr(dataset, store=None, mode='w-', synchronizer=None, group=None,
     _validate_dataset_names(dataset)
     _validate_attrs(dataset)
 
-    if mode == "a":
+    if (mode == 'a') or (append_dim is not None):
+        mode = 'a'
```

Comment:
Normally the preferred way to handle cases like this (intellligent default argument values) is to make them default to None. Thus, if you write `mode=‘w’` and `append_dim=‘time’`, you would get an error message rather than having your choice of mode overwritten.

Reply:
Thanks @shoyer, I've made mode default to None, which allows for checking inconsistency if append_dim is set and mode!='a' (I added a test for this). If mode and append_dim are not set, mode is internally set to `w-`, which is equivalent to the previous behavior.